### ISSUE TITLE
Truncating foreign referencing table causes deadlock

### ISSUE THREAD

**Issue by byucesoy** [citusdata/citus#938]
When we try to `TRUNCATE ... CASCADE` on foreign referenced table, it causes deadlock on worker nodes. That is because we send two separate `TRUNCATE ... CASCADE` command to workers, one for original table and one for referencing table. Each of these `TRUNCATE ... CASCADE` commands are sent via master_modify_multiple_shards UDF with separate transactions. They both obtain exclusive lock for referencing table. Thus waiting each other to commit.

Example;
```sql
-- Create two tables as;
SET citus.shard_replication_factor TO 1;

CREATE TABLE referenced_table(a int, b int, PRIMARY KEY(a));
SELECT create_distributed_table(referenced_table, 'a', 'hash');

CREATE TABLE referencing_table(a int, b int, FOREIGN KEY(a) REFERENCES referenced_table(a));
SELECT create_distributed_table(referencing_table, 'a', 'hash');

-- Then try to TRUNCATE referenced_table
-- This causes deadlock
TRUNCATE referenced_table CASCADE;
```

**Comment by byucesoy** [comment]
@sumedhpathak I think we should fix this for 6.0 release. Proper fix would be using transaction API but since it will not be ready for 6.0 release, we should come up with alternative solution.

**Comment by marcocitus** [comment]
I'm working on a fix for #898 that would also address this. It's basically a faster version of https://github.com/citusdata/citus/commit/9004852a1fb57cda5611d36e27c0df9d52111f57 , which doesn't have this problem.
